toffy/rename_fovs_test.py

```diff
@@ -0,0 +1,151 @@
+import tempfile
+import json
+import os
+import pytest
+import pandas as pd
+
+# from ark.utils import test_utils
+from ark.utils import io_utils
+from toffy import rename_fovs as rf
+
+
+def create_sample_run(data, create_json=False, bad=False):
+    fov_list = []
+    sample_run = {"fovs": fov_list}
+
+    # set up dictionary
+    for name, run_order, scan_count in \
+            zip(data.loc[:, "names"], data.loc[:, "run"], data.loc[:, "scans"]):
+        ex_fov = {
+            "scanCount": scan_count,
+            "runOrder": run_order,
+            "name": name
+        }
+        fov_list.append(ex_fov)
+
+    # delete name key if one is not provided
+    for fov in sample_run.get('fovs', ()):
+        if fov.get('name') is None:
+            del fov['name']
+
+    # create bad dictionary
+    if bad:
+        sample_run['bad key'] = sample_run['fovs']
+        del sample_run['fovs']
+
+    # create json file for the data
+    if create_json:
+        temp = tempfile.NamedTemporaryFile(mode="w", delete=False)
+        json.dump(sample_run, temp)
+        print(temp.name)
+        return temp.name
+
+    return sample_run
+
+
+def create_sample_fov_dirs(fovs, base_dir):
+    # make fov subdirectories
+    for fov in fovs:
+        os.mkdir(os.path.join(base_dir, fov))
+
+
+def remove_fov_dirs(base_dir):
+    # delete fov subdirectories
+    fovs = io_utils.list_folders(base_dir)
+    for fov in fovs:
+        os.rmdir(os.path.join(base_dir, fov))
+
+
+def test_check_unnamed_fovs():
+    # data with missing names
+    ex_data = pd.DataFrame(
+        {'names': ['MoQC', None, 'tonsil_bottom', 'moly_qc_tissue', None],
+         'run': list(range(1, 6)),
+         'scans': list(range(1, 6))
+         })
+
+    # create a dict with the sample data
+    ex_run = create_sample_run(ex_data)
+
+    # test that missing names are given a placeholder
+    rf.check_unnamed_fovs(ex_run)
+    for fov in ex_run.get('fovs', ()):
+        assert fov.get('name') is not None
+
+
+def test_rename_fov_dirs():
+    with tempfile.TemporaryDirectory() as base_dir:
+        # create run file and fov folder directories
+        dirs = ['run_folder', 'fov_folder']
+        for directory in dirs:
+            os.mkdir(os.path.join(base_dir, directory))
+        run_dir = os.path.join(base_dir, 'run_folder')
+        fov_dir = os.path.join(base_dir, 'fov_folder')
+
+        # create existing new directory
+        os.mkdir(os.path.join(base_dir, 'not_new_directory'))
+        not_new_dir = os.path.join(base_dir, 'not_new_directory')
+
+        # existing directory for new_dir should raise an error
+        with pytest.raises(ValueError, match="already exists"):
+            rf.rename_fov_dirs(run_dir, fov_dir, not_new_dir)
+
+        # regular sample run data
+        ex_data = pd.DataFrame(
+            {'names': ['MoQC', 'tonsil_bottom', 'moly_qc_tissue'],
```

Comment:
I would have the custom names also follow a pattern, like custom_1, custom_2, etc. Otherwise it makes it seem like these names specifically mean something